File: typescript/tests/executor.test.ts
/**
 * executor.ts — defensive dispatch over BOTH documented param shapes
 * (Reference §6 vs the local-automation section), coordinate scaling,
 * `raw` logged-and-skipped, terminal markers, and the NullBackend.
 */
import { describe, expect, it } from 'vitest';

import {
  ActionParamError,
  NullBackend,
  executeAction,
  executeActions,
} from '../src/coasty/executor.js';
import { type Action } from '../src/coasty/types.js';

function action(actionType: Action['action_type'], params: Record<string, unknown> = {}): Action {
  return { action_type: actionType, params };
}

describe('click', () => {
  it('clicks at {x, y} with left/1 defaults (Reference §6 shape)', async () => {
    const backend = new NullBackend();
    const result = await executeAction(action('click', { x: 100, y: 200 }), backend);
    expect(result).toEqual({ actionType: 'click', executed: true, terminal: null, detail: null });
    expect(backend.calls).toEqual([
      { method: 'click', args: [100, 200, { button: 'left', clicks: 1 }] },
    ]);
  });

  it('honors the optional button/clicks (local-automation shape)', async () => {
    const backend = new NullBackend();
    await executeAction(action('click', { x: 5, y: 6, button: 'right', clicks: 2 }), backend);
    expect(backend.calls[0]?.args[2]).toEqual({ button: 'right', clicks: 2 });
  });

  it('throws ActionParamError when coordinates are missing', async () => {
    await expect(executeAction(action('click', { x: 1 }), new NullBackend())).rejects.toThrow(
      ActionParamError,
    );
  });
});

describe('key_press — both shapes', () => {
  it('accepts {key: "enter"} (Reference §6)', async () => {
    const backend = new NullBackend();
    await executeAction(action('key_press', { key: 'enter' }), backend);
    expect(backend.calls).toEqual([{ method: 'keyPress', args: ['enter'] }]);
  });

  it('accepts {keys: [...]} pressed in order (local-automation)', async () => {
    const backend = new NullBackend();
    await executeAction(action('key_press', { keys: ['tab', 'enter'] }), backend);
    expect(backend.calls).toEqual([
      { method: 'keyPress', args: ['tab'] },
      { method: 'keyPress', args: ['enter'] },
    ]);
  });

  it('throws when neither key nor keys is present', async () => {
    await expect(executeAction(action('key_press'), new NullBackend())).rejects.toThrow(
      ActionParamError,
    );
  });
});

describe('key_combo', () => {
  it('presses the chord together', async () => {
    const backend = new NullBackend();
    await executeAction(action('key_combo', { keys: ['ctrl', 'c'] }), backend);
    expect(backend.calls).toEqual([{ method: 'keyCombo', args: [['ctrl', 'c']] }]);
  });
});

describe('wait — both shapes', () => {
  it('accepts {ms} (Reference §6)', async () => {
    const backend = new NullBackend();
    await executeAction(action('wait', { ms: 250 }), backend);
    expect(backend.calls).toEqual([{ method: 'wait', args: [250] }]);
  });

  it('accepts {seconds} and converts to ms (local-automation)', async () => {
    const backend = new NullBackend();
    await executeAction(action('wait', { seconds: 2 }), backend);
    expect(backend.calls).toEqual([{ method: 'wait', args: [2000] }]);
  });

  it('throws when neither ms nor seconds is present', async () => {
    await expect(executeAction(action('wait'), new NullBackend())).rejects.toThrow(
      ActionParamError,
    );
  });
});

describe('scroll — both shapes', () => {
  it('accepts {direction, amount} (Reference §6)', async () => {
    const backend = new NullBackend();
    await executeAction(action('scroll', { direction: 'down', amount: 3, x: 10, y: 20 }), backend);
    expect(backend.calls).toEqual([
      { method: 'scroll', args: [{ direction: 'down', amount: 3, x: 10, y: 20 }] },
    ]);
  });

  it('accepts signed {clicks}: positive scrolls up (local-automation)', async () => {
    const backend = new NullBackend();
    await executeAction(action('scroll', { clicks: 2 }), backend);
    expect(backend.calls).toEqual([
      { method: 'scroll', args: [{ direction: 'up', amount: 2, x: null, y: null }] },
    ]);
  });

  it('accepts signed {clicks}: negative scrolls down', async () => {
    const backend = new NullBackend();
    await executeAction(action('scroll', { clicks: -4 }), backend);
    expect(backend.calls).toEqual([
      { method: 'scroll', args: [{ direction: 'down', amount: 4, x: null, y: null }] },
    ]);
  });

  it('throws when neither shape is satisfied', async () => {
    await expect(
      executeAction(action('scroll', { direction: 'down' }), new NullBackend()),
    ).rejects.toThrow(ActionParamError);
    await expect(executeAction(action('scroll'), new NullBackend())).rejects.toThrow(
      ActionParamError,
    );
  });
});

describe('drag — both shapes', () => {
  it('accepts {from_x, from_y, to_x, to_y} (Reference §6)', async () => {
    const backend = new NullBackend();
    await executeAction(action('drag', { from_x: 1, from_y: 2, to_x: 3, to_y: 4 }), backend);
    expect(backend.calls).toEqual([{ method: 'drag', args: [1, 2, 3, 4] }]);
  });

  it('accepts {x1, y1, x2, y2} (local-automation)', async () => {
    const backend = new NullBackend();
    await executeAction(action('drag', { x1: 5, y1: 6, x2: 7, y2: 8 }), backend);
    expect(backend.calls).toEqual([{ method: 'drag', args: [5, 6, 7, 8] }]);
  });

  it('throws when endpoints are missing', async () => {
    await expect(
      executeAction(action('drag', { from_x: 1, from_y: 2 }), new NullBackend()),
    ).rejects.toThrow(ActionParamError);
  });
});

describe('move / type_text', () => {
  it('moves the pointer', async () => {
    const backend = new NullBackend();
    await executeAction(action('move', { x: 9, y: 8 }), backend);
    expect(backend.calls).toEqual([{ method: 'move', args: [9, 8] }]);
  });

  it('types text (preserving it exactly)', async () => {
    const backend = new NullBackend();
    await executeAction(action('type_text', { text: 'héllo  world!' }), backend);
    expect(backend.calls).toEqual([{ method: 'typeText', args: ['héllo  world!'] }]);
  });

  it('throws on a missing text param', async () => {
    await expect(executeAction(action('type_text'), new NullBackend())).rejects.toThrow(
      ActionParamError,
    );
  });
});

describe('coordinate scaling', () => {
  it('scales click/drag coordinates by (real / sent) and rounds', async () => {
    const backend = new NullBackend();
    // Sent a 1280x720 screenshot of a 1920x1080 screen -> scale 1.5x.
    const options = { scaleX: 1.5, scaleY: 1.5 };
    await executeAction(action('click', { x: 100, y: 33 }), backend, options);
    await executeAction(
      action('drag', { from_x: 10, from_y: 20, to_x: 30, to_y: 41 }),
      backend,
      options,
    );
    expect(backend.calls[0]?.args.slice(0, 2)).toEqual([150, 50]); // 33*1.5 = 49.5 -> 50
    expect(backend.calls[1]?.args).toEqual([15, 30, 45, 62]); // 41*1.5 = 61.5 -> 62
  });

  it('scales the optional scroll position but never the amount', async () => {
    const backend = new NullBackend();
    await executeAction(
      action('scroll', { direction: 'down', amount: 3, x: 100, y: 100 }),
      backend,
      { scaleX: 2, scaleY: 0.5 },
    );
    expect(backend.calls[0]?.args[0]).toEqual({ direction: 'down', amount: 3, x: 200, y: 50 });
  });
});

describe('raw — never executed', () => {
  it('logs and skips raw actions instead of executing them', async () => {
    const backend = new NullBackend();
    const logged: string[] = [];
    const result = await executeAction(
      action('raw', { code: 'pyautogui.hotkey("ctrl", "alt", "del")' }),
      backend,
      { logger: (message) => logged.push(message) },
    );

    expect(result.executed).toBe(false);
    expect(result.terminal).toBeNull();
    expect(result.detail).toContain('skipped raw action');
    expect(backend.calls).toEqual([]); // backend NEVER touched
    expect(logged).toHaveLength(1);
    expect(logged[0]).toContain('pyautogui.hotkey');
  });

  it('falls back to action.raw_code for the log preview', async () => {
    const logged: string[] = [];
    await executeAction(
      { action_type: 'raw', params: {}, raw_code: 'pyautogui.moveTo(1, 2)' },
      new NullBackend(),
      { logger: (message) => logged.push(message) },
    );
    expect(logged[0]).toContain('pyautogui.moveTo(1, 2)');
  });
});

describe('terminal markers', () => {
  it('done is terminal and executes nothing', async () => {
    const backend = new NullBackend();
    const result = await executeAction(action('done'), backend);
    expect(result).toEqual({ actionType: 'done', executed: false, terminal: 'done', detail: null });
    expect(backend.calls).toEqual([]);
  });

  it('fail is terminal and carries the reason', async () => {
    const result = await executeAction(
      action('fail', { reason: 'element not found' }),
      new NullBackend(),
    );
    expect(result.terminal).toBe('fail');
    expect(result.detail).toBe('element not found');
  });

  it('unknown action types throw ActionParamError', async () => {
    await expect(
      executeAction(
        { action_type: 'levitate' as Action['action_type'], params: {} },
        new NullBackend(),
      ),
    ).rejects.toThrow(ActionParamError);
  });
});

describe('executeActions', () => {
  it('executes in order and stops at the first terminal marker', async () => {
    const backend = new NullBackend();
    const results = await executeActions(
      [
        action('click', { x: 1, y: 2 }),
        action('type_text', { text: 'hi' }),
        action('done'),
        action('click', { x: 9, y: 9 }), // must NOT run
      ],
      backend,
    );

    expect(results).toHaveLength(3);
    expect(results[2]?.terminal).toBe('done');
    expect(backend.calls.map((c) => c.method)).toEqual(['click', 'typeText']);
  });
});

File: go/examples/internal/executor/executor_test.go
package executor

import (
	"bytes"
	"context"
	"strings"
	"testing"

	"github.com/coasty-ai/computer-use-cookbook/go/coasty"
)

func action(t coasty.ActionType, params map[string]any) coasty.Action {
	return coasty.Action{ActionType: t, Params: params}
}

// TestFormatActionBothShapes covers the canonical reference params AND the
// "local automation" alt shapes from the documented discrepancy.
func TestFormatActionBothShapes(t *testing.T) {
	identity := Scale{}
	double := NewScale(2560, 1440, 1280, 720)

	tests := []struct {
		name  string
		a     coasty.Action
		scale Scale
		want  string
	}{
		{
			name:  "click canonical identity",
			a:     action(coasty.ActionClick, map[string]any{"x": 100.0, "y": 50.0}),
			scale: identity,
			want:  "click at (100, 50)",
		},
		{
			name:  "click scaled",
			a:     action(coasty.ActionClick, map[string]any{"x": 100.0, "y": 50.0}),
			scale: double,
			want:  "click at (200, 100) [model (100, 50)]",
		},
		{
			name: "click alt shape with button and clicks",
			a: action(coasty.ActionClick, map[string]any{
				"x": 10.0, "y": 20.0, "button": "right", "clicks": 2.0,
			}),
			scale: identity,
			want:  "click at (10, 20) button=right clicks=2",
		},
		{
			name:  "move scaled",
			a:     action(coasty.ActionMove, map[string]any{"x": 640.0, "y": 360.0}),
			scale: double,
			want:  "move at (1280, 720) [model (640, 360)]",
		},
		{
			name:  "type_text",
			a:     action(coasty.ActionTypeText, map[string]any{"text": "hello"}),
			scale: identity,
			want:  `type_text "hello"`,
		},
		{
			name:  "key_press canonical key",
			a:     action(coasty.ActionKeyPress, map[string]any{"key": "enter"}),
			scale: identity,
			want:  `key_press "enter"`,
		},
		{
			name:  "key_press alt keys list",
			a:     action(coasty.ActionKeyPress, map[string]any{"keys": []any{"tab", "tab", "enter"}}),
			scale: identity,
			want:  "key_press sequence [tab tab enter]",
		},
		{
			name:  "key_combo",
			a:     action(coasty.ActionKeyCombo, map[string]any{"keys": []any{"ctrl", "c"}}),
			scale: identity,
			want:  "key_combo ctrl+c",
		},
		{
			name: "scroll canonical",
			a: action(coasty.ActionScroll, map[string]any{
				"x": 400.0, "y": 300.0, "direction": "down", "amount": 3.0,
			}),
			scale: identity,
			want:  "scroll down 3 at (400, 300)",
		},
		{
			name:  "scroll alt negative clicks means down",
			a:     action(coasty.ActionScroll, map[string]any{"clicks": -3.0}),
			scale: identity,
			want:  "scroll down 3",
		},
		{
			name:  "scroll alt positive clicks means up",
			a:     action(coasty.ActionScroll, map[string]any{"clicks": 2.0}),
			scale: identity,
			want:  "scroll up 2",
		},
		{
			name: "scroll alt clicks with explicit direction wins",
			a: action(coasty.ActionScroll, map[string]any{
				"clicks": 4.0, "direction": "left",
			}),
			scale: identity,
			want:  "scroll left 4",
		},
		{
			name: "drag canonical scaled",
			a: action(coasty.ActionDrag, map[string]any{
				"from_x": 10.0, "from_y": 20.0, "to_x": 30.0, "to_y": 40.0,
			}),
			scale: double,
			want:  "drag (20, 40) -> (60, 80) [model (10, 20) -> (30, 40)]",
		},
		{
			name: "drag alt shape",
			a: action(coasty.ActionDrag, map[string]any{
				"x1": 1.0, "y1": 2.0, "x2": 3.0, "y2": 4.0,
			}),
			scale: identity,
			want:  "drag (1, 2) -> (3, 4)",
		},
		{
			name:  "wait canonical ms",
			a:     action(coasty.ActionWait, map[string]any{"ms": 500.0}),
			scale: identity,
			want:  "wait 500ms",
		},
		{
			name:  "wait alt seconds",
			a:     action(coasty.ActionWait, map[string]any{"seconds": 1.5}),
			scale: identity,
			want:  "wait 1500ms",
		},
		{
			name:  "done",
			a:     action(coasty.ActionDone, nil),
			scale: identity,
			want:  "done — task complete, stop the loop",
		},
		{
			name:  "fail with reason",
			a:     action(coasty.ActionFail, map[string]any{"reason": "button is disabled"}),
			scale: identity,
			want:  "fail — agent reported failure: button is disabled",
		},
		{
			name:  "fail without reason",
			a:     action(coasty.ActionFail, nil),
			scale: identity,
			want:  "fail — agent reported failure (no reason given)",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := FormatAction(tt.a, tt.scale)
			if err != nil {
				t.Fatalf("FormatAction() error = %v", err)
			}
			if got != tt.want {
				t.Errorf("FormatAction() = %q, want %q", got, tt.want)
			}
		})
	}
}

func TestFormatActionRawIsNeverExecuted(t *testing.T) {
	a := action(coasty.ActionRaw, map[string]any{"code": "import os\nos.system('rm -rf /')"})
	got, err := FormatAction(a, Scale{})
	if err != nil {
		t.Fatalf("FormatAction() error = %v", err)
	}
	if !strings.Contains(got, "NOT executing") {
		t.Errorf("raw action line %q must state the code is not executed", got)
	}
	if !strings.Contains(got, "import os") || strings.Contains(got, "rm -rf") {
		t.Errorf("raw action line %q should show only the first line of code", got)
	}
}

func TestFormatActionUnknownTypeIsLoggedNotFatal(t *testing.T) {
	a := action(coasty.ActionType("hover"), map[string]any{"x": 1.0})
	got, err := FormatAction(a, Scale{})
	if err != nil {
		t.Fatalf("FormatAction() error = %v (unknown types must not fail the loop)", err)
	}
	if !strings.Contains(got, `unknown action type "hover"`) {
		t.Errorf("FormatAction() = %q, want a logged unknown-type notice", got)
	}
}

func TestFormatActionMissingParams(t *testing.T) {
	tests := []struct {
		name string
		a    coasty.Action
	}{
		{"click without y", action(coasty.ActionClick, map[string]any{"x": 1.0})},
		{"type_text without text", action(coasty.ActionTypeText, nil)},
		{"key_press neither key nor keys", action(coasty.ActionKeyPress, nil)},
		{"key_combo without keys", action(coasty.ActionKeyCombo, nil)},
		{"scroll neither amount nor clicks", action(coasty.ActionScroll, map[string]any{"x": 1.0})},
		{"scroll amount without direction", action(coasty.ActionScroll, map[string]any{"amount": 2.0})},
		{"drag partial coords", action(coasty.ActionDrag, map[string]any{"from_x": 1.0, "from_y": 2.0, "to_x": 3.0})},
		{"wait neither ms nor seconds", action(coasty.ActionWait, nil)},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if _, err := FormatAction(tt.a, Scale{}); err == nil {
				t.Errorf("FormatAction(%v) expected a missing-params error", tt.a)
			}
		})
	}
}

func TestScale(t *testing.T) {
	s := NewScale(2560, 1440, 1280, 720)
	if x, y := s.Apply(640, 360); x != 1280 || y != 720 {
		t.Errorf("Apply(640,360) = (%d,%d), want (1280,720)", x, y)
	}
	if s.Identity() {
		t.Error("2x scale must not be identity")
	}
	var zero Scale
	if x, y := zero.Apply(33, 44); x != 33 || y != 44 {
		t.Errorf("zero Scale must be identity, got (%d,%d)", x, y)
	}
	if !zero.Identity() {
		t.Error("zero Scale.Identity() = false, want true")
	}
	if degenerate := NewScale(0, -1, 1280, 720); !degenerate.Identity() {
		t.Error("degenerate real dims must fall back to identity")
	}
}

func TestLoggingExecute(t *testing.T) {
	var buf bytes.Buffer
	exec := &Logging{W: &buf, Scale: NewScale(200, 100, 100, 50), Prefix: "step 1 "}
	a := coasty.Action{
		ActionType:  coasty.ActionClick,
		Params:      map[string]any{"x": 50.0, "y": 25.0},
		Description: "Click the submit button",
	}
	if err := exec.Execute(context.Background(), a); err != nil {
		t.Fatalf("Execute() error = %v", err)
	}
	got := buf.String()
	want := "step 1 [exec] click at (100, 50) [model (50, 25)] — Click the submit button\n"
	if got != want {
		t.Errorf("Execute() wrote %q, want %q", got, want)
	}

	if err := exec.Execute(context.Background(), action(coasty.ActionClick, nil)); err == nil {
		t.Error("Execute() with missing params must return an error")
	}
}
